cumulusci/core/flowrunner.py

```diff
@@ -581,3 +586,96 @@ def _find_result_by_path(self, path):
             if result.path[-len(path) :] == path:
                 return result
         raise NameError("Path not found: {}".format(path))
+
+
+class PreflightFlowCoordinator(FlowCoordinator):
+    """Coordinates running preflight checks instead of the actual flow steps.
+    """
+
+    def run(self, org_config):
+        self.org_config = org_config
+        self.callbacks.pre_flow(self)
+
+        self._init_org()
+        self._rule(fill="-")
+        self.logger.info("Organization:")
+        self.logger.info("  {}: {}".format("Username", org_config.username))
+        self.logger.info("  {}: {}".format("  Org Id", org_config.org_id))
+        self._rule(fill="-", new_line=True)
+
+        self.logger.info("Running preflight checks...")
+        self._rule(new_line=True)
+
+        self.jinja2_context = {
+            "tasks": TaskCache(self, self.project_config, org_config),
+            "project_config": self.project_config,
+            "org_config": self.org_config,
+        }
+
+        self.preflight_results = defaultdict(list)
+        try:
+            # flow-level checks
+            for check in self.flow_config.checks or []:
+                result = self.evaluate_check(check)
+                if result:
+                    self.preflight_results[None].append(result)
+
+            # Step-level checks
+            for step in self.steps:
+                for check in step.task_config.get("checks", []):
+                    result = self.evaluate_check(check)
+                    if result:
+                        self.preflight_results[step.path].append(result)
+        finally:
+            self.callbacks.post_flow(self)
+
+    def evaluate_check(self, check):
+        self.logger.info("Evaluating check: {}".format(check["when"]))
+        expr = jinja2_env.compile_expression(check["when"])
+        value = bool(expr(**self.jinja2_context))
+        self.logger.info("Check result: {}".format(value))
+        if value:
+            return {"status": check["action"], "message": check.get("message")}
+
+
+class TaskCache(object):
+    """Provides access to named tasks and caches their results.
+
+    This is intended for use in a jinja2 expression context
+    so that multiple expressions evaluated in the same context
+    can avoid running a task more than once with the same options.
+    """
+
+    def __init__(self, flow, project_config, org_config):
```

Comment:
Is there an intent behind passing these values directly instead of accessing them through `flow`?

Reply:
No, I forgot they're available on the flow. Thanks. I'll refactor this.